File: 055_jump_game/Solution.go
package main

import "fmt"

func canJump(nums []int) bool {
    arrLen := len(nums)
    rightMost := 0

    for i := 0; i < arrLen; i++ {
        if i <= rightMost {
            rightMost = max(rightMost, i + nums[i])
            if rightMost >= arrLen - 1 {
                return true
            }
        }
    }

    return false
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func main() {
    nums := []int{2, 3, 1, 1, 4}
    fmt.Println(canJump(nums))
}

File: 055_jump_game/Solution.py
class Solution:
    def canJump(self, nums: list) -> bool:
        right_most = 0
        arr_len = len(nums)

        for i in range(arr_len):
            if i <= right_most:
                right_most = max(right_most, i + nums[i])
                if right_most > arr_len - 1: return True

        return False


if __name__ == '__main__':
    nums = [2, 3, 1, 1, 4]
    print(Solution().canJump(nums))

